### ISSUE TITLE
Number Of Lookups Statistics

### ISSUE THREAD

**Issue by neilord** [themoeway/yomitan#741]
It would be very cool to have the daily number of lookup stats! We can use it to see how much we have immersed and how better overall we are getting. It could potentially be a new tab "Statistics" on the settings page.

**Comment by danielxt** [comment]
Hey [👋](https://emojipedia.org/waving-hand), I'm interested in implementing this.

I would create a new tab in the settings page like you said:

![image](https://github.com/themoeway/yomitan/assets/114246026/89e36e3b-3bdb-4a22-97b8-2622ca66752a)

Then I would represent daily lookup statistics using a heatmap across the last year. I'll use [Cal Heatmap](https://cal-heatmap.com/) to generate the heatmap.

![image](https://github.com/themoeway/yomitan/assets/114246026/1897ec1d-305b-4a31-b790-33353ca81db0)

I would also like to add more things to the Statistics tab so I would be open to more suggestions.

**Comment by jamesmaa** [comment]
I'm open to this suggestion. Let's implement this iteratively and start with agreeing on how/what format we want to store the data and have a very simple statistic

**Comment by danielxt** [comment]
[Cal-Heatmap can take in a json file](https://cal-heatmap.com/docs/options/data) as the data input. From checking out ext/schemas/, I can see that this project uses json format to store data so I will use the same to keep it consistent. The keys will be the date and the value will be the number of lookups like so:
`{ date: '2012-01-01', value: 3 },
  { date: '2012-01-02', value: 6 },`
I would create a new json file under ext/schemas/ called lookup-statistics.json and store lookup data in there.

**Comment by Neilord** [comment]
Hey @jamesmaa and @danielxt,

Great that you are willing to implement this!
One of the most important parts is creating a proper data schema for the statistics tab. After having the data collected we will be able to generate whatever charts we like.

I was thinking that we could add something like this to some `local.storage` statistics object every time the word hovers:
```json
{
  "word": "the hovered word",
  "date": "date with time probably",
  "popup-level": "the current level of the popup, whether it is a popup created after hovering the context of the page or another popup",
}
```
and update it with something like this when the popup is closed:
```json
{
  "time-hovering": "The time popup was open",
  "last-dictionary": "until what dictionary was popup scrolled before closed",
}
```

Then really useful stats could be those:
* Number of popup hovers a day
* Average number of popup hovers before learning a word (learned word could be a word that has not hovered in the last [everange number of hours using yomitan a day * 30]
* Most popular `last-dictionary`

**Comment by danielxt** [comment]
Sounds good. I would make the Statistics tab like this then:

![image](https://github.com/themoeway/yomitan/assets/114246026/f0ff85e5-0f41-4d3e-b97f-8e0944cfb041)

**Comment by danielxt** [comment]
Hi all, I'm having some trouble importing cal-heatmap into Yomitan. I did `npm install cal-heatmap` and I set up a heatmap-controller.js to handle all my main logic in ext/js/pages/settings. I then tried `import CalHeatmap from 'cal-heatmap'` in that and got a type error. 

I noticed that for the already installed dependencies like dexie, parse5, and wanakana, modules were first exported in dev/lib/ and then referenced in the ext/js/pages/settings/ file where needed. backup-controller.js for instance backs all the way up and then gets to the export file with `'../../../lib/dexie.js'`. I then tried to follow this pattern by creating a cal-heatmap.js export file in dev/lib/ and doing `import {CalHeatmap} from '../../../../dev/lib/cal-heatmap.js';` but now I just get file not found errors.

TLDR: what's the process to install and use a new external module in this project?

**Comment by Neilord** [comment]
@danielxt, thank you for keeping the development of those features!

I have not been working with the Yomitan codebase, so can't give you advice on how to install new modules.
But would suggest you postpone the heatmap development if it takes a too long time (despite it is 100% useful feature).

It could be better to just start with implementing simple stats, make a pull request, and then if it is approved more developers will be able to commit to the stats implementation.

No one likes huge pull requests and we better start with checking if Yomitan devs will accept that stats pull request at all before it will take you a lot of time. :)

**Comment by jamesmaa** [comment]
I agree with @Neilord. I think we should be very careful with introducing new modules (security, licensing, etc) and the core devs (namely toasted-nutbread) hasn't been around so if you're hoping to get a PR with a new module in, it's going to take much longer.

**Comment by jamesmaa** [comment]
Hey @Neilord and @danielxt sorry for the rugpull I chatted with the other yomitan maintainers and we've decided that this feature is actually not very aligned with yomitan's core purpose so we're not going to merge this PR into master.
